cfg_mgmt/github.py

```diff
@@ -0,0 +1,326 @@
+        cfg_element=cfg_element,
+        cfg_factory=cfg_factory,
+    )
+
+    if len(local_sources) != 1:
+        raise NotImplementedError(
+            'Do not know how to rotate github configs sourced from more than one file.'
+        )
+
+    source_file = next((s for s in local_sources))
+
+    known_github_configs: typing.Iterable[GithubConfig] = [
+        c for c in cfg_factory._cfg_elements(
+            cfg_type_name=cfg_element._type_name,
+        )
+    ]
+
+    cfg_name = cfg_element.name()
+    if not (cfg_to_rotate := next(
+        filter(lambda c: c.name() == cfg_name, known_github_configs),
+        None,
+    )):
+        raise RuntimeError(
+            f"Did not find requested config '{cfg_name}' in config dir at '{cfg_dir}'"
+        )
+
+    old_keys = [{
+        'name': u.username(),
+        'public_key': update_user(cfg_to_rotate, u.username()),
+    } for u in cfg_to_rotate._technical_users()]
+
+    _write_github_configs(
+        cfg_dir=cfg_dir,
+        cfg_file_name=source_file,
+        github_configs=known_github_configs,
+    )
+
+    _update_config_queue(
+        cfg_dir=cfg_dir,
+        queue_file_name=cfg_queue_fname,
+        config_element=cfg_to_rotate,
+        config_name=cfg_name,
+        config_queue=cfg_queue,
+        secret_id={'github_users': old_keys},
+    )
+
+    _update_config_status(
+        cfg_dir=cfg_dir,
+        cfg_status_filename=cfg_status_fname,
+        config_element=cfg_to_rotate,
+        config_statuses=cfg_metadata.statuses,
+    )
+
+    try:
+        _push_to_github(
+            cfg_dir=cfg_dir,
+            config_element=cfg_to_rotate,
+            git_helper=git_helper,
+            target_ref=target_ref,
+        )
+    except:
+        logger.info('Undoing changes in GitHub ...')
+        for u in cfg_to_rotate._technical_users():
+            undo_update(github_config=cfg_to_rotate, username=u.username())
+        raise
+
+
+def _write_github_configs(cfg_dir, cfg_file_name, github_configs):
+    configs = {c.name(): c.raw for c in github_configs}
+    with open(os.path.join(cfg_dir, cfg_file_name), 'w') as cfg_file:
+        yaml.dump(configs, cfg_file, Dumper=MultilineYamlDumper)
+
+
+def _read_config_queue(cfg_dir, queue_file_name) -> dict:
+    return ci.util.parse_yaml_file(os.path.join(cfg_dir, queue_file_name))
+
+
+def _write_config_queue(cfg_dir, queue_file_name, config_queue):
+    with open(os.path.join(cfg_dir, queue_file_name), 'w') as queue_file:
+        yaml.dump(config_queue, queue_file, Dumper=MultilineYamlDumper)
+
+
+def _pretty_commit_message(config_element):
+    commit_message = (
+        f"Rotate element of config type '{config_element._type_name}'\n\n"
+
+        f'changed element: {config_element.name()}'
+    )
+    if ci.util._running_on_ci():
+        commit_message += f'\nJob url: {concourse.util.own_running_build_url}'
+
+    return commit_message
+
+
+def _push_to_github(
```

Comment:
the name does not match the semantics -> better: create_commit_and_push?
also, in theory, the push-target would not have to be a github-hosted git repository (just omit gh?)